Updating correlation surface docs for nondeterminism (#881)
tqec/tqec: docs/user_guide/terminology.rst

You can think of constructing the correlation surface as moving a line of logical operators through the structure,
only allowing the logical operators to attach to walls with the same basis.
The physical qubit measurements and stabilizer measurements in the correlation surface determine the sign relationship between the logical operators at the input and output. Any required Pauli operator corrections are tracked in a classical data structure called a Pauli frame. The tracking does not delay the circuit, unless there is an operation which needs the correct Pauli frame in real time.

> Related concepts

Author: Dealing with this stuff is kind of annoying, is there an automated way to enforce these formatting requirements?

> The physical qubit measurements and stabilizer measurements in the correlation surface determine the sign relationship between the logical operators at the input and output. Any required Pauli operator corrections are tracked in a classical data structure called the :ref:`Pauli frame <pauli_frame>`. The tracking does not delay the circuit, unless there is an operation which needs the correct Pauli frame in real time.

Author: Moved to Pauli frame section and changed correlation surface section to be about producing a PPM rather than a Pauli correction.

~~~~~~~~~~~~~~~~

> A set of measurements with predictable parity in the absence of errors is called a detector. The detectors which inform the Pauli frame of a logical observable constitute a correlation surface. The detecting regions highlighted in Crumble and annotated in Stim are a labeling of the spacetime stabilizers manifested by detectors at a physical circuit level.

Author: suggested change applied

> Two quantum computations are logically equivalent if they both implement the same set of correlation surfaces. The process of rewriting a block graph into one that has the same correlation surfaces is essentially a topological deformation, because maintaining the connectivity between logical blocks is necessary and sufficient for preserving the list of correlation surfaces. The stabilizer ZX calculus is a mathematically rigorous diagrammatic language for reasoning about these deformations; correlation surfaces roughly correspond to Pauli webs in the ZX calculus. The correspondence between ``tqec`` block graphs and ZX graphs is sufficiently accurate for ``tqec`` to use ZX graphs as an intermediate representation, but one may find subtle differences in terms depending on the class of ZX graphs one is analyzing.

Reviewer: ```suggestion
Two Clifford quantum computations are logically equivalent if they both implement the same set of correlation surfaces. The process of rewriting a block graph into one that has the same correlation surfaces is essentially a topological deformation, because maintaining the connectivity between logical blocks is necessary and sufficient for preserving the list of correlation surfaces. The stabilizer ZX calculus is a mathematically rigorous diagrammatic language for reasoning about these deformations; correlation surfaces roughly correspond to Pauli webs in the ZX calculus. The correspondence between ``tqec`` block graphs and ZX graphs is sufficiently accurate for ``tqec`` to use ZX graphs as an intermediate representation, but one may find subtle differences in terms depending on the class of ZX graphs one is analyzing.
```

Should emphasize Clifford computation. When non-Clifford gate comes into play, two computation might be not equivalent even they have the same correlation surfaces.

> Two quantum computations are logically equivalent if they both implement the same set of correlation surfaces. The process of rewriting a block graph into one that has the same correlation surfaces is essentially a topological deformation, because maintaining the connectivity between logical blocks is necessary and sufficient for preserving the list of correlation surfaces. The stabilizer ZX calculus is a mathematically rigorous diagrammatic language for reasoning about these deformations; correlation surfaces roughly correspond to Pauli webs in the ZX calculus. The correspondence between ``tqec`` block graphs and ZX graphs is sufficiently accurate for ``tqec`` to use ZX graphs as an intermediate representation, but one may find subtle differences in terms depending on the class of ZX graphs one is analyzing.

Author: > Maintaining connectivity alone is not sufficient; the boundary basis matches (the color rules) also need to be preserved.

Ended up deleting the sentence. I think block graph rewrite semantics will require a separate section of documentation. 

> I think it's formally called the phase-free ZX calculus? Not sure, could be a slightly different thing.

Okay I looked into this. Although the phase free ZX calculus captures the surface code (and all CSS codes), this is only in the sense of their encoding maps. Our diagrams include T state initialization and Y basis initialization and measurement which are nodes of phase pi/4 and pi/2 respectively. My new writing states that blocks graphs are an instantiation of the Clifford+T fragment, meaning all nodes are integer multiples of pi/4.

Although Clifford+T fragment is the most accurate classification, this fragment seems to lack less provable guarantees than the stabilizer fragment (which is just integer multiples of pi/2 nodes, (excludes T nodes)). In practice, ``tqec`` uses the rewrite rules of the stabilizer ZX calculus. Because of the equivalences in this [this paper](https://arxiv.org/pdf/2603.09580), specifically section 4, I don't expect our introduction of pi/4 nodes to break anything, especially because they will always be in the form of a teleportation gadget whose flow propagation rules are known and able-to-be-defined independently of the ZX calculus.  

> Formally ZX diagram.

Added an explicit definition of a ZX graph. 

> We can cite https://arxiv.org/abs/1704.08670 if we want.

Done. Also added a bunch of other citations from the ZX calculus community, mainly because I want readers interested in formal support for the IR to refer to those results.

> Should emphasize Clifford computation. When non-Clifford gate comes into play, two computation might be not equivalent even they have the same correlation surfaces.

Fixed, also included Austin's clarification below.

> In particular, since post-selection is not a physically-realistic substitute for measurement, the ``tqec`` compiler views measurement operations that are not explicitly paired with a classical feedforward instruction by the user as either :math:`T` gates or a "discard" indicating that the measurement outcome has no effect on the program.

Author: The motivation for writing this came from some confusion I experienced when reviewing David's `pyzx` parser. I have added some explicit references that came up during that conversation which cleared my doubts. 

Post-selection formally means something stronger and less physical than measurement and feedforward. It means you measure the qubit, and pretend somehow that you only keep the branch where the measurement result was the one you wanted. That is not a realistic replacement for actual measurement control, because in hardware you cannot force the desired outcome; you can only observe the outcome and adapt, or discard failed runs.

The key point I a want to add to the docs is that measurement-dependent corrections must be explicit. Otherwise, the compiler assumes the measurement has no program-level classical consequence, rather than silently assuming an idealized post-selected branch. 

I expect this terminology and logic to not change even after we build proper control system infrastructure to ``tqec``.

> It is possible for a block graph to support a logical observable that is non-deterministic. This occurs when the detectors which support the logical observable, despite potentially maintaining its fault tolerance, do not have deterministic parity in the absence of errors. In other words, the data provided by the observable is random. For example, consider a surface code patch initialized in the :math:`Z` basis and then measured in the :math:`X` basis. Tracing the :math:`X` observable back to a :math:`Z` initialization would specify a totally random event. Generally speaking, this makes it difficult to tell whether the computation was fault tolerant or not, because there is no deterministic value that could serve as a ground truth. For this reason, the ``tqec`` compiler avoids tracing correlation surfaces corresponding to non-deterministic observables, and raises an error when no deterministic correlation surfaces are found.

Reviewer: ```suggestion
It is possible for a block graph to support a logical observable that is non-deterministic. This occurs when the measurements which support the logical observable do not have deterministic parity, even in the absence of errors. For example, consider a surface code patch initialized in the :math:`Z` basis and then measured in the :math:`X` basis. Tracing the :math:`X` observable back to a :math:`Z` initialization would specify a totally random event. Generally speaking, this makes it difficult to tell whether the computation was fault tolerant or not, because there is no deterministic value that could serve as a ground truth. For this reason, the ``tqec`` compiler avoids tracing correlation surfaces corresponding to non-deterministic observables, and raises an error when no deterministic correlation surfaces are found.
```

Author: Exact change applied.

> It is possible for a block graph to support a logical observable that is non-deterministic. This occurs when the detectors which support the logical observable, despite potentially maintaining its fault tolerance, do not have deterministic parity in the absence of errors. In other words, the data provided by the observable is random. For example, consider a surface code patch initialized in the :math:`Z` basis and then measured in the :math:`X` basis. Tracing the :math:`X` observable back to a :math:`Z` initialization would specify a totally random event. Generally speaking, this makes it difficult to tell whether the computation was fault tolerant or not, because there is no deterministic value that could serve as a ground truth. For this reason, the ``tqec`` compiler avoids tracing correlation surfaces corresponding to non-deterministic observables, and raises an error when no deterministic correlation surfaces are found.

Author: Exact change applied.

      Quantum circuit measuring the ``XX`` stabilizer.

.. _detector:

Detector
--------

A detector is a set of one or more measurements that are
supposed to have a deterministic parity in the absence of errors. :footcite:`McEwen_2023`

.. _pauli_frame:

> Pauli Frame

Reviewer 2: Can cite https://www.nature.com/articles/nature03350 for Pauli frame, if we want.

Author: Done

-----------

The Pauli frame is a classical data structure that, in each execution of a quantum program,
stores the effect of the Pauli operations that were determined to be necessary by the decoder
and program specification. If there were no logical errors during circuit execution, the decoding
system updates the Pauli frame with the value of the Pauli operator that corrects the error which

> the decoding algorithm outputs as the most likely explanation for each detection event. A detection

Author: Thanks, exact change applied.

      Quantum circuit measuring the ``XX`` stabilizer.

.. _detector:

Detector
--------

A detector is a set of one or more measurements that are
supposed to have a deterministic parity in the absence of errors :footcite:`McEwen_2023`.

.. _pauli_frame:

Pauli Frame
-----------

The Pauli frame is a classical data structure that, in each execution of a quantum program,
stores the effect of the Pauli operations that were determined to be necessary by the decoder

> and program specification :footcite:`Knill_2005`. The Pauli correction is given by the parity of measurements on :ref:`correlation surfaces <correlation_surface>`. The parity is directly flipped when decoding. The tracking only delays the circuit if an operation which needs the correct Pauli frame is scheduled in a blocking manner.

Reviewer 2: Sorry, my previous comment was inaccurate. Even with windowed decoding, the boundary condition being passed is not exactly a physical Pauli frame. Fortunately Austin clarified this (that we don't need a physical Pauli frame). As for the logical Pauli frame, I don't think it's needed if there's no Pauli in the logical circuit (or we apply them physically instead of virtually) and we track correlation surfaces instead of byproduct operators. That said, if you do have a logical Pauli frame, I believe it should be applied after computing the correlation surface parity.

      Quantum circuit measuring the ``XX`` stabilizer.

.. _detector:

Detector
--------

A detector is a set of one or more measurements that are
supposed to have a deterministic parity in the absence of errors :footcite:`McEwen_2023`.

.. _pauli_frame:

Pauli Frame
-----------

The Pauli frame is a classical data structure that, in each execution of a quantum program,
stores the effect of the Pauli operations that were determined to be necessary by the decoder

> and program specification :footcite:`Knill_2005`. The tracking does not delay the circuit, unless there is an operation which needs the correct Pauli frame in real time. If there were no logical errors during circuit execution, the decoding system updates the Pauli frame with the value of the Pauli operator that corrects the error which the decoding algorithm outputs as the most likely explanation for the observed set of detection events. A detection event is a deviation in the expected parity of a :ref:`detector <detector>`.

Reviewer 2: > If there were no logical errors during circuit execution

Seems unnecessary? Logical errors are unknown in real execution, and the Pauli frame will be updated with the decoder's prediction regardless.
